openedx/openedx-platform
# Implement third party auth ID mapping API

**UPDATE Feb.16, 2016**
Since "client_credentials" grant type is supported know, there is no need to manually create authorization code. The workflow below reflects this change.

**Description:**
This implements a new API that can be used to retrieve a list of mappings between third party auth remote IDs and edX usernames. For use cases and more details, please see [proposal + discussion](https://docs.google.com/document/d/1Uifq_uDRfPUzba7mikWo_F_S89ynDwkQ6D0-SXhXl9g/edit?usp=sharing).
- This PR is based on @bradenmacdonald 's PR https://github.com/edx/edx-platform/pull/9672. Large part of the code has been reviewed there.
- The current authentication for this API is using either EDX_API_KEY or OAuth2 Authorization.

**Background:**
When an edX course is used in an on-campus environment and SAML, LTI or other SSO is enabled to allow students to authenticate to edX using their on-campus credentials, normally there will be an identifier sent from an on-campus IdP. This identifier is stored in edX and linked to the edX account. Later when students return from an on-campus IdP, edX is able to authenticate the student using their edX account. 

Because students may create edX accounts not using their real identity, the instructors will have difficulty to identify the edX students in their courses and won't be able to upload grades from edX back to on-campus gradebook system as the downloaded grades from edX only contain edX IDs, which are unknown to on-campus system.

This new API can be used for mapping between the edX user ID and the ID provided by identity provider (IdP). This API will give information that will allow instructors or course support staff to figure out the identity of the on-campus students who are using the edX course. It will also allow grade information coming from edX to be appropriately associated to campus students for uploading to an on-campus learning management system (LMS) or student information system (SIS). 

This API is suppose to be consumed by a on-campus middleware, which combines the information from on-campus system. The middleware would talk to campus system to get information about course and enrolment. Then use those information to control instructor access. So the instructor will be able to get his/her edX student on-campus identity, uploading grades back to their LMS or SIS, making groups based on on-campus activities, etc.

This API also helps with troubleshooting with the issue where students put in wrong email addresses and weren't able to activate their accounts. Thus they weren't able to login EdX through Shibboleth because of the inactive accounts. With this API, on-campus course support staff or instructor can provide EdX support the student real EdX username instead of the ID that Shibboleth passed over to EdX, thus avoid the EdX support manually querying the tables to figure out the mapping between Shibboleth ID and EdX username.

**JIRA:** N/A

**Component Affected:** LMS (Third Party Auth)

**Users Affected:** Currently users with EDX_API_KEY (EdX internal) or OAuth2 client authorization (external use)

**Timeframe:** Mid Oct.

**Reviewers:** @bradenmacdonald and TBD

**Resources Needed:** None

**Database Table Changes:**
- This PR adds a new table called third_party_auth_apipermissions. So a `paver update_db` is required after merging.

**Test Instructions:**
- Enable one or more third party auth providers (such as 'Dummy') and link them to user's accounts.
- Setup authentication and authorization:
  - For using EDX_API_KEY: Edit ~/lms.auth.json and set EDX_API_KEY to some value, e.g. api-key. And restart LMS.
  - For using OAuth2:
    - Setup an OAuth2 Client. In Django Admin, go to OAuth2 -> Client -> Add Client. 
      - User:  it is good idea to create a dedicated user for this API
      - URL, Redirect URL: http://localhost// (not being used in this API)
      - Client Type: Confidential
      - Other fields: no specific requirements
    - ~~Create an authorization code. As EdX/django-auth2-provider doesn't support client_credential grant type, we need to create the authorization code manually. Go to OAuth2 -> Grants -> Add grant. Remember the code for later use.~~
    - Give the new client permission to this API. Go to Third_Party_Auth -> Provider API Permissions -> Add Provider API Permission
      - Select the new client and the provider created in the previous step. This will give the new client access to query that provider mapping.
- Test this API with commands like below, where {provider_id} should be replaced with the actual ID. If it is SAML provider, provider ID should be "saml-PROVIDER_SLUG", e.g., saml-ubc. For OAuth2 providers, it will be the "oa2-BACKEND_NAME", e.g., oa2-google
  - Using EDX_API_KEY
    
    ``` bash
    curl --header "X-edx-API-key: api-key" http://localhost:8000/api/third_party_auth/v0/providers/{provider_id}/users 
    
    curl --header "X-edx-API-key: api-key" http://localhost:8000/api/third_party_auth/v0/providers/{provider_id}/users?username=USERNAME1,USERNAME2
    
    curl --header "X-edx-API-key: api-key" http://localhost:8000/api/third_party_auth/v0/providers/{provider_id}/users?username=USERNAME1&username=USERNAME2
    
    curl --header "X-edx-API-key: api-key" http://localhost:8000/api/third_party_auth/v0/providers/{provider_id}/users?remote_id=REMOTE_ID1,REMOTE_ID2
    
    curl --header "X-edx-API-key: api-key" http://localhost:8000/api/third_party_auth/v0/providers/{provider_id}/users?remote_id=REMOTE_ID1&remote_idREMOTE_ID2
    ```
  - Using OAuth2
    - ~~Using the authorization code created in the previous step to get the access token~~
    
    ``` bash
    curl --data "grant_type=authorization_code&code=CODE&client_id=CLIENT_ID&client_secret=CLIENT_SECRET" http://localhost:8000/oauth2/access_token
    Return: {"access_token": "2983fbbc6c41947a2807946678d594b06b403dbf", "token_type": "Bearer", "expires_in": 31535999, "refresh_token": "909c2b0926d642283f762f7437c1a9cd12844b07", "scope": ""}
    ```
    - Using client credential to get the access token
    
    ``` bash
    curl --data "client_id=CLIENT_ID&client_secret=CLIENT_SECRET&grant_type=client_credentials" http://localhost:8000/oauth2/access_token
    Return: {"access_token": "c1efde84445b2f256e1c80886b3f6d46339b84ee", "token_type": "Bearer", "expires_in": 31535999, "scope": ""}
    ```
    - Using access token to issue request to API:
    
    ``` bash
    curl --header "Authorization: Bearer ACCESS_TOKEN" http://localhost:8000//api/third_party_auth/v0/providers/{provider_id}/users
    
    curl --header "Authorization: Bearer ACCESS_TOKEN" http://localhost:8000//api/third_party_auth/v0/providers/{provider_id}/users?username=USERNAME1,USERNAME2
    
    curl --header "Authorization: Bearer ACCESS_TOKEN" http://localhost:8000//api/third_party_auth/v0/providers/{provider_id}/users?username=USERNAME1&username=USERNAME2
    
    curl --header "Authorization: Bearer ACCESS_TOKEN" http://localhost:8000//api/third_party_auth/v0/providers/{provider_id}/users?remote_id=REMOTE_ID1,REMOTE_ID2
    
    curl --header "Authorization: Bearer ACCESS_TOKEN" http://localhost:8000//api/third_party_auth/v0/providers/{provider_id}/users?remote_id=REMOTE_ID1&remote_id=REMOTE_ID2
    ```
- API should return a json object including the mappings. E.g.:

``` json
{
  "page": 1,
  "page_size": 200,
  "count": 8,
  "results": [
    {"username": "USERNAME1", "remote_id": "REMOTE_ID1"},
    {"username": "USERNAME2", "remote_id": "REMOTE_ID2"},
  ]
}
```
